### ISSUE TITLE
CFE should use environment to resolve conditional imports and check against interface files

### ISSUE THREAD

**Issue by sivachandra** [dart-lang/sdk#30143]
When I compile using the `fasta` tool, I see that only the default import of a conditional import is picked up. A simple way to see this is to compile the following code with `fasta`:

```
import 'package:resource/resource.dart';

void main() {
  Resource r;
}
```

The package `resource` uses conditional imports in one of its source files. However, IR dump of the generated `.dill` file shows that the default import was picked up instead of the correct conditional import.

The [conditional import proposal](https://github.com/munificent/dep-interface-libraries/blob/master/Proposal.md) says that the conditions are tested using _environment declarations_. So, the `fasta` tool and the front-end API should probably be extended to take optional _environment declarations_.

`dartanalyzer` supports specifying environment variables via the `-D` command line option but I am not sure if these are like the above _environment declarations_.

@sigmundch @peter-ahe-google

**Comment by matanlurey** [comment]
This is an area I'd like to kept in the loop how implementation will work.

(Especially regarding bazel, where we pre-build ahead of time per package)

**Comment by peter-ahe-google** [comment]
Conditional imports basically break modular compilation and code editing. As long as that is the case, it's a very low priority for us.

**Comment by mraleph** [comment]
We do need conditional imports to work correctly if we want to switch to Fasta, because people do use these language features or use packages that use these features and some of our own package use conditional imports (see for example: https://github.com/flutter/flutter/issues/9413).

**Comment by peter-ahe-google** [comment]
I think we need to escalate this to the language team. There's no way to support conditional imports in the current Kernel format without breaking modular compilation.

**Comment by eernstg** [comment]
A conditional import that makes a choice among several alternatives based on information which is available at compile time should not break modular compilation: If library L was compiled with environment E then the transitive closure of imports is determined by E, and that compiled version of L would be valid for linking with anything that also receives environment E.

So there is a need to keep a separate copy of the generated code for each configuration. But that is obviously necessary when the conditional import mechanism does not make any guarantees about similarities of alternatives (in particular, they are not guaranteed to be similar to the default).

Is that "no way"?

**Comment by peter-ahe-google** [comment]
You're assuming that the environment is bound at compile time. That's precisely what breaks modular compilation and makes it hard to implement code editing of this feature.

Ideally, the environment is bound at link time, but we have no way to represent that in Kernel, and it still doesn't help the code editing part of this feature.

My experience from google3 is that changing compiler flags is problematic, because it requires you to generate code for each configuration, and because two modules with different compiler options can't be used together.

**Comment by eernstg** [comment]
Suppose we have this:
```dart
... // Some conditional imports.
class C extends B {
  var x = y;
}
```
then `y` may be declared at the global scope of a conditionally imported library in one configuration, and it may be a member of `B` (to which this occurrence of `y` is resolved because there is no global `y`) in another configuration. I think it will be at least inconvenient to generate code for all possible variants and then later (at link time, when the configuration is known) to select the appropriate one. I believe that, in practice, the configuration _must_ be available at compile time.

**Comment by peter-ahe-google** [comment]
I'm not disagreeing. I'm pointing out that the specification is forcing us to break modular compilation. The specification could be based on first-class libraries (or library interfaces), and it wouldn't require us to break modular compilation.

An even simpler possibility is just using what dart2js has been implementing for years: dummy implementations of platform libraries that aren't supported in a given configuration.

Oh, did I say specification. Sorry, I meant proposal. We don't even have a specification for this feature, right?

**Comment by matanlurey** [comment]
I have a document internally for landing this feature with Bazel (go/dart-config-imports), but the crux of this is adding a new `platform` attribute that can be applied at the Bazel _target_ level:

```build
# package:collection
dart_library(
  name = "collection",
  # No platform =, since we don't use configurable imports.
)
```

```build
load("...", "WEB_PLATFORM", "VM_PLATFORM")

# package:resource
dart_library(
  name = "resource",
  # Has a configurable import for either dart:html or dart:io.
  platform = [WEB_PLATFORM, VM_PLATFORM],
)
```

In this case `package:collection` is compiled once, because it doesn't need any platform-specific compilation, and `package:resource` is compiled twice - once for `dart:io`-supporting compilers and once for `dart:html`. Anyway, the full "specification" is in the doc internally.

---

I imagine the same logic holds externally for modular compilation, that is:

1. For a full program compile (no caching), you can just set the environment flag
2. For modular compilation, you need to follow something like _platform = [ ... ]_  per package
3. For IDEs or the analyzer, you always use the default implementation

That being said, I realize Dart does not have any sort of standard way of defining that information (afaik, `pubspec.yaml` is entirely a pub thing, and is unused by other tooling), so I'm actually in agreement with @peter-ahe-google that this needs to be respecified. I'm actually a fan of re-thinking it with explicit _platform_ support anyway, instead of based on arbitrary environment variables representing if core libraries should exist or not.

**Comment by eernstg** [comment]
Right, it's not specified.

However, the notion of first-class libraries or library interfaces has not been part of the proposal or its implementation for a long time. It is essentially a textual include of an alternative, and there is no limit on the number or nature of the differences among different alternatives.

Also, it is a bad idea for IDEs to rely on the default: There is no reason to believe that any static analysis property based on any particular alternative (e.g., the default) will be valid for any other alternative (types may be different, a given type name may have different members, lookups may be resolved in different scopes, etc). The sane approach is to specify the configuration in the IDE, and then do work based on that particular configuration.

In order to achieve genuine separate compilation we would need library interfaces, as Peter mentions, and each alternative would then have to conform to the declared interface for a specific configurable import. We have proposals for very simple ones ("level one": only values can be configurably-imported, not types), but there was a strong desire for being able to extend/implement configurable imported classes, which would require a much more elaborate notion of library interfaces.

**Comment by matanlurey** [comment]
> Also, it is a bad idea for IDEs to rely on the default: There is no reason to believe that any static analysis property based on any particular alternative (e.g., the default) will be valid for any other alternative (types may be different, a given type name may have different members, lookups may be resolved in different scopes, etc). The sane approach is to specify the configuration in the IDE, and then do work based on that particular configuration.

Just an FYI, this *isn't* what we do today, the analyzer always uses the default configuration. I realize this is potentially error-prone for the reasons you mention, but this language feature never considered how any of this would work 😜 

> In order to achieve genuine separate compilation we would need library interfaces, as Peter mentions, and each alternative would then have to conform to the declared interface for a specific configurable import. We have proposals for very simple ones ("level one": only values can be configurably-imported, not types), but there was a strong desire for being able to extend/implement configurable imported classes, which would require a much more elaborate notion of library interfaces.

Then it sounds like we should toss all of this from 2.0 (/cc @dgrove)

**Comment by eernstg** [comment]
> the analyzer always uses the default configuration

I know, but I still think that we _should_ strive for a sane approach, even if it isn't so convenient.

**Comment by peter-ahe-google** [comment]
The sane approach is *not* to for the IDE to use a particular configuration. The sane approach is to design the feature so it isn't necessary for a developer to run through static analysis of all the possible configurations in which a library can be used.

**Comment by eernstg** [comment]
That's a trade off. I think it makes sanse to iterate over all configurations when they may differ in ways that we cannot predict, and it makes sense to enforce that all configurations are similar and then analyze the software for all configurations once and for all; but it is not a good idea to mislead developers to think that they've checked all configurations when they haven't.

**Comment by sigmundch** [comment]
I believe that Peter was suggesting that the language could ensure that the default and other implementations statically equivalent, and the user wouldn't have to worry about looking at things outside of the default implementation.

Regardless of how that is resolved, we have a problem with modular compilation because config-imports are configured with environment variables, which are defined globally. We need a way to specify environment values consistently in a modular fashion (In fact, we have this problem even if we ignore config-imports).

I believe we can create a modular environment variable system though: we can require users to provide the values on each modular build and check for consistent values. That is, if a modular build depends on a summary/outline compiled with different environment values, then the compiler could provide an compile-time error.

Just wanted to add a small detail to Matan's description of our plan in bazel: In build systems like bazel, we want to avoid a plethora of combinations of environment values, so we want to go further and restrict config-imports to only allow environment values that are defined explicitly by a platform (and exclude user-provided values). E.g. `platform = vm` defines `dart.library.io`. This makes the integration with the build system more manageable and easier to understand. I personally would like this restriction to be in the language, since the main purpose of config-imports is to choose implementation details based on the platform.

**Comment by matanlurey** [comment]
Yup, thanks for mentioning that @sigmundch.

This also comes into play with how we'd _like_ to get `pub [serve|build]` working and would also potentially impact how you'd have `flutter` CLI work. For example, if I have the following `pubspec.yaml`:

```yaml
name: my_package

dependencies:
  collection:
```

You should have _all_ the information you need, at the point of `pub [get|upgrade|downgrade]` to _fetch_ the package, and _pre-compile_ it. Here, _pre-compile_ would mean compiling it to Kernel:

* `pub get` detects `package:collection` needs an update
* `package:collection` is fetched either from cache or network
* `package:collection` is pre-compiled to the Kernel format and stored in `.pub`

Now subsequent rebuilds for either development or release of `my_package` don't need to anything about `package:collection` - it has already been "pre-compiled" to a format that we can directly give to the VM/DDC/etc.

For configurable imports to work in this scenario, we have a similar issue to bazel.

**Comment by sivachandra** [comment]
For my own education, what does "modular compilation" exactly mean?

**Comment by matanlurey** [comment]
It means that if you have the following dependency tree:

`dart:ui` --> `package:flutter` --> `package:flutter_gallery`

It is possible to compile them in isolation, and store the results in a cache, but still have a hermetic build process every time you say `flutter run flutter_gallery`. I don't know the specifics about the kernel, but you can think of it the following way I assume:

```
var dartUiKernel = compileOrUseCache('dart:ui');
var flutterKernel = compileOrUseCache('package:flutter');
var flutterGalleryKernel = compileOrUseCache('package:flutter_gallery');
runApplication('package:flutter_gallery', deps: [
  dartUiKernel,
  flutterKernel,
  flutterGalleryKernel,
]);
```

In a modular compilation pipeline, we can build (and cache) both `dart:ui` and `package:flutter` on disk, never touching it (or re-compiling) and re-using the artifacts of that compilation process in any application (such as `flutter_gallery`) that needs it.

**Comment by sigmundch** [comment]
For us modular compilation implies a few things: 
 - it's a mechanism to compile libraries piecemeal: we can group libraries in "build units" and create .dill files for each build unit as a separate invocation of the compiler.
 - it's a mechanism to compile libraries without knowing what program will run them - so we can't assume anything about how those libraries are used
 - Note that we don't imply "separate compilation": we expect to build an outline of dependencies before we compile a library/build-unit. These outlines are explicit inputs to the compiler.

This is the same underlying mechanism we use today to build platform.dill separate from the rest of the app. Using Matan's example, we can compile `dart:ui` in isolation, but we need the outline of 'dart:ui' to compile `package:flutter`. 

At some point an app needs to run, so there will be a link step. 

The reason modular-compilation comes up with config-imports is that there is general agreement that the compiler should generate code for a specific configuration and not keep around code for all configurations. Otherwise we risk adding too much extra complexity in the IR and in all tools that consume the IR.

**Comment by sivachandra** [comment]
Thanks for the detailed explanation @matanlurey and @sigmundch.

**Comment by eernstg** [comment]
> I believe that Peter was suggesting that the language could ensure that the default and
> other implementations statically equivalent, and the user wouldn't have to worry about
> looking at things outside of the default implementation.

Right, that was also my interpretation of Peter's words, and I said the same thing in my [answer to Peter](https://github.com/dart-lang/sdk/issues/30143#issuecomment-314795372):

> to achieve genuine separate compilation we would need library interfaces, as Peter mentions,
> and each alternative would then have to conform to the declared interface for a specific
> configurable import

In this case, the default import would play two roles: It specifies a library interface, and it provides an implementation. The other alternatives would then be checked against that library interface.

However, the mechanism that we have on the table now does not support any such thing: There are no constraints on the kinds of dependencies that an importer may have on a conditionally imported library, and we would need substantial extensions to the static analysis of Dart in order to support anything that gets even close to this level of coupling (or lack of encapsulation, if you wish).

For instance, we'd need a concept of "the type of a class" which is sufficiently detailed to allow (1) specifying the properties of a given configurably imported class `C` in so much detail that we would be able to (2) check that any given class `C` in the j'th alternative conforms to that class-type, and (3) check that usages like `class D extends C..` and `foo(C x) => ... x ...`, etc. are correct, and (4) detect the properties of any such configuration dependent entity. (NB: such an entity may be _anywhere_ in your software, as long as it transitively imports anything through a configurable import). As an example of a type of information needed for this which is not needed in the current static analysis, we'd have to know that the given configurably imported class does _not_ declare a getter named `foo` when checking that it's OK for `D` to declare a method named `foo` (normal class related types do not specify that a class does _not_ have certain things). Similarly, in order to check `T foo(S s) {..}` in `D`, the class-type of `C` needs to constrain which declarations of `foo` are OK for `C` to have, if any, and we should note that `C` could inherit such a `foo`, and different alternatives could (indeed, would!) have different imports; and even if such a `foo` accepts an argument of type `S`, `S` might mean different things in different alternatives.

As a trivial example of a regime that would work: Let L_D be the default library for a given configurable import. The library interface is defined to be I := erase_body(L_D), where erase_body is a function that deletes all function bodies (and possibly field initializers and similar things, assuming that inference has already been performed and all inferred types added). To check that any given library L conforms to this library interface, check that erase_body(L) == I, that is, they are textually identical. Crucially, they must also have exactly the same imports.

As you can see, this is trivial, easy to implement, and useless. In order to obtain something which is actually useful we must allow different alternatives to differ, and that's where we need to be careful in order to maintain correctness. We have to be more precise than "they are very similar".

So we have the current design: Allow all dependencies and check per configuration. Easy to do, obviously correct, but tedious when there are many configurations.

Alternatively, we could have a notion of library interfaces with conformance checking, and we do have proposals for simple versions of this. This would allow true separation of configuration dependent and configuration independent parts of the software (hence separate compilation). But it's less simple.

**Comment by peter-ahe-google** [comment]
> So we have the current design: Allow all dependencies and check per configuration. Easy to do, obviously correct, but tedious when there are many configurations.

It's not easy to do: checking all configurations isn't a feasible solution which is the point I've been making from the start of this conversation. The number of configurations are unbounded and likely to become huge, as was the case for GWT. This feature needs to be redesigned, as there's no reason to repeat mistakes of the past.

In GWT, the number of configurations on some projects were: number of supported browsers times number of supported OSes times number of localization times number of experiments and this led to multiple hour compile times.

The number of supported browsers could be something like 10. IE10, IE11, Chrome, Safari, Firefox, ...

The number of supported OSes could be something like 10. Mac, Linux, Windows, Android, iOS, tvOS, ChromeOS, Fuchsia, ...

The number of localizations could be [50](https://support.google.com/googleplay/android-developer/table/4419860?hl=en)

I have about 10 experiments in gmail right now if I just count the labs, but I'm guessing that there might be even more so let's say 20.

10 * 10 * 50 * 20 is 100000

That's a lot of configurations to analyze while the user is waiting for a code completion dialog to pop up.

**Comment by eernstg** [comment]
> checking all configurations isn't a feasible solution .. there might be .. 100000

Sure, the number of configurations can explode (Sigmund suggested that we could limit the selection to support only platform defined environment variables, but even that could yield many combinations). So the near-textual-include mechanism that we have requires a strong discipline, or a small project.

I'm just saying that if we want library interfaces then we should _have_ them, we shouldn't just pretend.

**Comment by peter-ahe-google** [comment]
> I'm just saying that if we want library interfaces then we should have them, we shouldn't just pretend.

If that is indeed what you're saying, you need to work on your delivery. Because what I've been hearing is that you've been telling me how to implement this feature and saying that it's easy.

I'm strongly in favor library interfaces, but I'm trying to stick to providing feedback on what isn't working. It's possible that there are other alternatives, but the feature as it currently stands isn't feasible.

> So the near-textual-include mechanism that we have requires a strong discipline, or a small project.

That's not a feasible solution. This boils down to doing the same thing as GWT, but *hoping* that we don't end up having the same problems.

**Comment by eernstg** [comment]
> what I've been hearing is that you've been telling me how to implement this feature and saying
> that it's easy.

I proposed from the very beginning how we could have a simple version of library interfaces with well-understood static checking and clean separation of configuration dependent and configuration independent libraries: Values (e.g., top-level functions) can be imported configurably, types can not. That model, however, is not powerful enough to allow inheritance from a configurably imported class.

I certainly never claimed that it would be easy to define a notion of library interfaces that would allow anything like the freedom of the currently implemented (per-configuration) mechanism to depend on every detail of a configurably imported library. Obviously, I haven't been claiming that it's easy to implement something that I've argued we are far from being able to specify.

So you like library interfaces, and I like library interfaces, but it does not help to pretend that the current mechanism has them, or even that it can get them in any reasonable amount of time.

That said, I agree that a per-configuration approach has serious scalability issues, but I do not think that it is useless. In particular, it is _extremely_ flexible about how you can use configurably imported entities.

**Comment by peter-ahe-google** [comment]
> I proposed from the very beginning how we could have a simple version of library interfaces with well-understood static checking and clean separation of configuration dependent and configuration independent libraries: Values (e.g., top-level functions) can be imported configurably, types can not.

How is it relevant to this bug what you proposed from the very beginning? This bug is not an introspective into the discussions that you may or may not already have had. This bug is about the concrete problem that we have now: we can't implement a feature as it is specified.

> I certainly never claimed that it would be easy to define a notion of library interfaces that would allow anything like the freedom of the currently implemented (per-configuration) mechanism to depend on every detail of a configurably imported library. 

I never said that you did. What I did say was:

> It's not easy to do

In response to you saying:

> So we have the current design: Allow all dependencies and check per configuration. Easy to do

> So you like library interfaces, and I like library interfaces, but it does not help to pretend that the current mechanism has them, or even that it can get them in any reasonable amount of time.

I'm not pretending anythin. I'm saying that the current specification isn't feasible to implement, and I specifically said that I'm only trying to provide feedback about what isn't working, not how to fix it.

> That said, I agree that a per-configuration approach has serious scalability issues, but I do not think that it is useless. In particular, it is extremely flexible about how you can use configurably imported entities.

We're not debating if the feature is useless. We're debating whether or not it is feasible to implement.

In addition, we have concrete evidence from GWT that the feature as it currently stands leads to severe problems. That doesn't mean that the feature isn't useful, it means that we should be extremely careful and examine other possibilities.

**Comment by eernstg** [comment]
I responded to things you said. E.g., you said to me 'you need to work on your delivery', which must mean 'write a spec for library interfaces', and I responded that we've had such one for a simple case for a long time, and we won't get one which is as flexible as the per-configuration approach. But let's take this IRL, I want to learn where things get difficult for you.

**Comment by peter-ahe-google** [comment]
I meant work on delivery as in how you deliver a message. I did not mean deliverable or specification.

**Comment by peter-ahe-google** [comment]
I think we have a simple solution that will unblock Flutter @sivachandra

**Comment by sivachandra** [comment]
Is the "simple solution" buried somewhere in the discussion here? Flutter is not blocked the this issue per say as we are not using the front-end yet. The issue which actually prompted me to file this bug did end up having a simple solution: https://github.com/flutter/flutter/commit/dba29aa0fce684b201c6b03efc73ba607c7b507b

**Comment by peter-ahe-google** [comment]
It's late here, and I'll add more details tomorrow.

**Comment by sigmundch** [comment]
The "simple" solution is that we will start by supporting constants of the form `dart.library.*`, which are directly derived from the platform, but we'll not support arbitrary environment values at this time.  This subset of the feature is aligned with the plan we have for modular support in bazel as well.

Later, once we finalize the language spec on this feature so it can always work modularly, we can come back to it and finalize it in front-end as well.

**Comment by peter-ahe-google** [comment]
What Siggi said :-)

**Comment by sivachandra** [comment]
I think the VM also does the same today.

**Comment by a-siva** [comment]
Has the simple solution ("we will start by supporting constants of the form dart.library.*") landed ?

**Comment by keertip** [comment]
Cannot run any of the flutter tests with front end internally, as the flutter test runner uses a conditional import.

**Comment by peter-ahe-google** [comment]
@mraleph can you confirm that you're blocked on this to be able to run flutter?

**Comment by kmillikin** [comment]
@sjindel-google will implement the simple solution of supporting boolean constants of the form `dart.library.*` and configuring them appropriately for each back-end target.  Do you (@keertip, @a-siva ) know of a simple way that we can get a list of the constants that we need to support for the VM target?

**Comment by sjindel-google** [comment]
These revisions introduced partial support for conditional imports to CFE:
https://dart-review.googlesource.com/c/sdk/+/34721
https://dart-review.googlesource.com/c/sdk/+/34940

This level of support should be enough for Flutter and "Bazel".
There are still two non-conforming aspects of the implementation:

* The conditional import selected is used for both static analysis and as the execution dependency. The informal spec. for conditional imports requires that the interface (default) import is used for static analysis, and that the CFE verifies that the implementation file adheres to the interface file.

* Only import conditions of the form `dart.library.*` are handled. Other import conditions should be evaluated by reference to command-line defines. This will introduce complications with modular compilation.

**Comment by keertip** [comment]
@sjindel-google, still seeing the same failure with tests, not able to resolve conditional imports. Opened an issue in internal bug tracker with more details.

**Comment by zanderso** [comment]
@rmacnak-google FYI

**Comment by rmacnak-google** [comment]
@kmillikin See Api::GetEnvironmentValue in runtime/vm/dart_api_impl.cc for the constants defined by the VM:

dart.library.{core, async, io, ...}
dart.vm
dart.vm.product
dart.developer.causal_async_stacks

**Comment by sjindel-google** [comment]
The problem here was that I implemented conditional imports, but actually conditional exports were needed.

I'll work on that very soon.

**Comment by rmacnak-google** [comment]
Blocks https://fuchsia-review.googlesource.com/c/topaz/+/118419

**Comment by sjindel-google** [comment]
Should the priority be changed?

**Comment by zanderso** [comment]
I don't know if the priority should be increased, but the sooner there is a fix for this, the sooner we can start trying out the CFE in Fuchsia's build, which is the CL that @rmacnak-google linked.

**Comment by rmacnak-google** [comment]
Ta!

**Comment by zanderso** [comment]
Thanks!
